README.md
# Nemo-IAP iOS SDK (version 2.1.1)

**This guide shows you how to integrate your iOS app using the NemoSDK for iOS. The NemoSDK for iOS consists of the following component SDKs:**
  - [The NemoSDK Core includes in-app purchases](https://github.com/gosusdk/ios-nemosdk_iap_demo/releases)
  - Nemo OpenId: [AppAuth.framework](https://github.com/gosusdk/ios-nemosdk_iap_demo/releases)
  - Grpc framework
  - Appsflyer framework
  - Firebase framework

#### FEATURES:
  - [Nemo Login: Authenticate people with NemoID credentials.](#integrate-nemosdk)
  - [Nemo in-app purchase](#integrate-nemosdk-iap)
  - Nemo Tracking: 
    * [Tracking with AppsFlyer](#integrate-nemosdktracking-appsflyer)
    * [Tracking with Firebase](#integrate-nemosdktracking-firebase)

<a name="integrate-nemosdk"></a>
## Integrate NemoSDK with in-app purchase

- Embed NemoSDK latest version and Third party framework into your project
- Some other libraries: 
  - AppAuth.framework
  - absl.framework
  - grpc.framework
  - GRPCClient.framework
  - openssl_grpc.framework
  - Protobuf.framework
  - ProtoRPC.framework
  - RxLibrary.framework
  - uv.framework

#### 1. Configure NemoSDK in your project (default info.plist)
  ```xml
    <key>CFBundleURLTypes</key>
    <array>
      <dict>
        <key>CFBundleURLSchemes</key>
        <array>
          <string>[RedirectURL/EndSessionEndpoint]</string>
        </array>
      </dict>
    </array>
    <key>ClientID</key>
    <string>[ClientID]</string>
    <key>AuthorizationScope</key>
    <string>openid email phone_number profile offline_access</string>
    <key>Issuer</key>
    <string>[Issuer]</string>
    <key>GameClientID</key>
    <string>[GameClientID]</string>
    <key>SdkSignature</key>
    <string>[SdkSignature]</string>
  ```
- **Issuer**: The fully qualified issuer URL of the server (example: https://gid-uat.nemoverse.io)
- **[RedirectURL/EndSessionEndpoint]**: URL Schemes (example: nemo.app.demo.app)
- **GameClientID, SdkSignature:** will provide for each product
  
#### 2. AppAuth Framework Embed
![photo_2022-11-23_11-38-37](https://user-images.githubusercontent.com/94542020/203470313-a5eed93b-1e10-43cd-bee2-bf95c4bd5768.jpg)

#### 3. Initialize NemoSDK
```objectivec
//AppDelegate.h
#import "NemoSDK/NemoSDK.h"
@interface AppDelegate : UIResponder <UIApplicationDelegate>
@end

//AppDelegate.m
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // Override point for customization after application launch.
    [[NemoSDK sharedInstance] sdkInit];
    return YES;
}
```
#### 4. NemoSDK - Authorization Interface
```objectivec
//MainViewController.m
#import <NemoSDK/NemoSDK.h>
@interface MainViewController ()<LoginDelegate,IAPDelegate>
@end

@implementation MainViewController
  //return onLoginSuccess/onLoginFailure delegate
  [[NemoSDK sharedInstance] login];

  //return json string
  [[NemoSDK sharedInstance] getUserInfo]

  //use as onLogoutFailure/onLogoutSuccess Delegate
  [[NemoSDK sharedInstance] logout];

  #pragma Login Delegate

  - (void)onLoginFailure:(NSString *)message {
      NSLog(@"onLoginFailure = %@", message);
  }

  - (void)onLoginSuccess:(NSString *)access_token andIdToken:(NSString *)id_token {
      NSLog(@"onLoginSuccess = %@ - %@", access_token, id_token);
  }

  - (void)onLogoutFailure:(NSString *)message {
      NSLog(@"onLogoutFailure = %@", message);
  }

  - (void)onLogoutSuccess:(NSString *)message {
      NSLog(@"onLogoutSuccess = %@", message);
  }
```
<a name="integrate-nemosdk-iap"></a>
#### 5. In-app purchase
```objectivec
  //MainViewController.m
  #import <NemoSDK/NemoSDK.h>
  @interface MainViewController ()<LoginDelegate,IAPDelegate>
  @end
  
  - (void) call_iap
  {
    NSString *productID = @"vn.devapp.pack1";
    NSString *appleSecret = @"";
    NSString *orderID = @"";
    NSString *orderInfo = @"Item 300 gold";
    NSString *amount = @"22000";
    NSString *server = @"22";
    NSString *character = @"Character_Name";
    NSString *extraInfo = @"Extra_Info";
    NSString *userInfoString = [[NemoSDK sharedInstance] getUserInfo];
    NSJSONSerialization *userData = NULL;
    if([userInfoString length] > 0) {
        userData = [NSJSONSerialization JSONObjectWithData:[userInfoString dataUsingEncoding:NSUTF8StringEncoding] options:0 error:nil];
    }
    NSString *username = [userData valueForKey:@"sub"];
    
    IAPDataRequest *iapData = [[IAPDataRequest alloc] initWithData:username andOrderId:orderID andOrderInfo:orderInfo andServerID:server andAmount:amount andAppleProductID:productID andAppleShareSecrect:appleSecret andRoleID:character andExtraInfo:extraInfo];
    
    [[NemoSDK sharedInstance] showIAP:iapData andMainView:self andIAPDelegate:self];
  }

  #prama IAP delegate
  - (void) IAPInitFailed:(NSString *)message andErrorCode:(NSString *)errorCode {}
  - (void) IAPPurchaseFailed:(NSString *)message andErrorCode:(NSString *)errorCode {}
  - (void) IAPCompleted:(NSString *)message {}
```

<a name="integrate-nemosdktracking-appsflyer"></a>
## Integrate NemoSDKTracking
### I. Tracking with Appsflyer
- Embed NemoSDKTracking latest version and Third party framework into your project
- Some other libraries: [AppsFlyerLib.framework](https://github.com/itcgosucorp/nemosdk-ios/releases)

##### 1. Configure NemoSignIn in your project (default info.plist)
```xml    
    <key>AppsFlyerAppleID</key>
    <string>0123456789</string>
    <key>AppsFlyerKey</key>
    <string>AppsFlyerKey000001111</string>
    <key>NSUserTrackingUsageDescription</key>
    <string>This identifier will be used to deliver personalized ads to you.</string>
  ```
##### 2. AppsFlyerLib Framework Linker
<img width="475" alt="image" src="https://user-images.githubusercontent.com/94542020/207208331-8eb41bd6-a3f2-4f58-b122-700b881951bc.png">

##### 3. Initialize NemoSDKTracking
```objectivec
//AppDelegate.m
#import "NemoSDKTracking/NemoSDKTracking.h"

- (void)applicationDidBecomeActive:(UIApplication *)application {
    [[NemoSDKTracking sharedInstance] initTracking:application];
    application.applicationIconBadgeNumber = 0;
}
```

##### 4. Function Interface
```objectivec
[[NemoSDKTracking AppsFlyer] trackingEventLoginOnAF:@"userId" andAccount:@"account"];
[[NemoSDKTracking AppsFlyer] trackingEventOnAF:@"event_abc" withValues:@{
    @"key1": @"account1",
    @"key2": @"account2"
}];
[[NemoSDKTracking AppsFlyer] trackingLevelArchiveEventOnAF:@"userId" andAccount:@"account" andLevel:@"12301"];
```

<a name="integrate-nemosdktracking-firebase"></a>
### II. Tracking with Firebase
- Embed NemoSDKTracking latest version and Third party framework into your project
- Some other libraries: [Firebase SDK](https://github.com/itcgosucorp/nemosdk-ios/releases)

##### 1. Configure NemoSignIn in your project (default info.plist)
- Move GoogleService-Info.plist file into the root of your Xcode project. If prompted, select to add the config file to all targets.
- Insert -ObjC to “Other Linker Flags ”on Xcode Project: Main target -> build settings -> search "other linker flags"
- Configure Tracking Usage Description into .plist file (default: info.plist)*.
  Open with source and insert code: 
  ```xml
  <key>NSUserTrackingUsageDescription</key>
  <string>This identifier will be used to deliver personalized ads to you.</string>
  ```
  
##### 2. Firebase Framework Linker
<img width="473" alt="image" src="https://user-images.githubusercontent.com/94542020/207208287-654585c2-8892-4227-9da0-7176f3ddce81.png">

##### 3. Initialize NemoSDKTracking with Firebase
```objectivec
//AppDelegate.m
#import "NemoSDKTracking/NemoSDKTracking.h"

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    //...
    [[NemoSDKTracking sharedInstance] applicationDelegate:self andApplication:application didFinishLaunchingWithOptions:launchOptions];
    return YES;
}
```

##### 4. Function Interface
```objectivec
[[NemoSDKTracking Firebase] trackingEventOnFirebase:@"eventName" parameters:@{@"eventEventLogKey":@"eventEventLogValue"}];
[[NemoSDKTracking Firebase] trackingScreenOnFirebase:@"screenName" screenClass:@"screenClass"];
[[NemoSDKTracking Firebase] setUserPropertiesOnFirebase:@"userValue" forName:@"usernameName"];
```

By using the NemoSDK for iOS you agree to these terms.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

### Changed
## [v2.0.2](https://github.com/itcgosucorp/nemosdk-ios/releases/tag/v2.0.2)
#### NemoSDKTracking framework
-  Add Tracking with Firebase

## [v2.0.1](https://github.com/itcgosucorp/nemosdk-ios/releases/tag/v2.0.1)
#### NemoSDK Framework
##### Configure NemoSignIn in your project (default info.plist)
1. NemoScope key replaced by AuthorizationScope
2. NemoUrl key replaced by Issuer

##### Initialize NemoSDK
```objectivec
//AppDelegate.m
[[NemoSDK sharedInstance] sdkInit:nil];
//replaced by
[[NemoSDK sharedInstance] sdkInit];
[NemoSDK sharedInstance].loginDelegate = self;
```

##### Authorization Interface
```objectivec
[[NemoSDK sharedInstance] login:self andDelegate:self];
[[NemoSDK sharedInstance] logoutBackground:self andDelegate:self];
//replaced by
[[NemoSDK sharedInstance] login];
[[NemoSDK sharedInstance] logout];
```
#### NemoSDKTracking framework
-  Add Tracking with AppsFlyer
And, Add Tracking functionality to the SDK. [See more](https://github.com/itcgosucorp/nemosdk-ios#integrate-nemosdktracking)


NemoSDKDemo/Classess/Controller/MainViewController.m

#define IS_IPAD (UI_USER_INTERFACE_IDIOM() != UIUserInterfaceIdiomPhone)
#define IS_IPHONE (!IS_IPAD)

#import "MainViewController.h"
#import <NemoSDK/NemoSDK.h>
#import <NemoSDK/SdkConfig.h>
#import <NemoSDKTracking/NemoSDKTracking.h>


@interface MainViewController ()<LoginDelegate,IAPDelegate>

@end

@implementation MainViewController

#pragma Login Delegate

- (void)onLoginFailure:(NSString *)message {
    NSLog(@"onLoginFailure = %@", message);
//    [[SdkHelper sharedInstance] showAlertMessage:self andWithTitle:@"onLoginFailure" andWithMessage:message andCallback:nil];
}

- (void)onLoginSuccess:(NSString *)access_token andIdToken:(NSString *)id_token {
    NSLog(@"onLoginSuccess = %@ - %@", access_token, id_token);
    [[NemoSDKTracking AppsFlyer] trackingEventLoginOnAF:@"userId" andAccount:@"neropro"];
    [[NemoSDKTracking AppsFlyer] trackingEventOnAF:@"event_abc" withValues:@{
        @"key1": @"nero1",
        @"key2": @"nero2"
    }];
    [[NemoSDKTracking AppsFlyer] trackingLevelArchiveEventOnAF:@"userId" andAccount:@"nerocasten" andLevel:@"12301"];
    _btn_showiap.hidden = NO;
    _btn_userinfo.hidden = NO;
    _btn_logout.hidden = NO;
    _btn_signin.hidden = YES;
}

- (void) onLogoutSuccess:(NSString *)message
{
    NSLog(@"onLogoutSuccess = %@", message);
//    _btn_showiap.hidden = YES;
    _btn_userinfo.hidden = YES;
    _btn_logout.hidden = YES;
    _btn_signin.hidden = NO;
}

- (IBAction)btnSignIn:(id)sender {
    [[NemoSDK sharedInstance] login:self];
}


- (IBAction)btnUserInfo:(id)sender {
    NSLog(@"userInfo = %@", [[NemoSDK sharedInstance] getUserInfo]);
    NSLog(@"refreshToken = %@", [[NemoSDK sharedInstance] getRefreshToken]);
    [self callTrackingFirebaseExample];
    if([[NemoSDK sharedInstance] getUserInfo]) {
        NSLog(@"show abc");
    }
}
- (IBAction)btnLogout:(id)sender {
    [[NemoSDK sharedInstance] logout];
}
- (IBAction)btnClickShowIAP:(id)sender {
    NSString *productID = @"vn.devapp.pack1";
    /* password from apple share secret (chuỗi thông tin khi tạo các gói Automatically Renewable Subscription) - dùng để verify iap, trường hợp không phải gói Automatically Renewable Subscription thì để trống */
    NSString *appleSecret = @"";
    //orderID was gernarate from server game and only one (mã hoá đơn là duy nhất không được trùng)
    NSString *orderID = [NSString stringWithFormat:@"%i_10095276_2_13_11", (int)[[NSDate date] timeIntervalSince1970]];
    //description package will purchase (mô tả gói cần mua)
    NSString *orderInfo = @"Item 300 gold";
    //amount
    NSString *amount = @"22000";//this is amount of money (VNĐ)
    //server
    NSString *server = @"22";
    NSString *character = @"Character_Name";
    NSString *extraInfo = @"Extra_Info";
    NSString *userInfoString = [[NemoSDK sharedInstance] getUserInfo];
    NSJSONSerialization *userData = NULL;
    if([userInfoString length] > 0) {
        userData = [NSJSONSerialization JSONObjectWithData:[userInfoString dataUsingEncoding:NSUTF8StringEncoding] options:0 error:nil];
    }
    if(!userData) {
//        return;
    }
    NSString *username = [userData valueForKey:@"sub"];
    orderID = @"";
    //Show log value of parameters
    NSLog(@"product_id:%@, apple_secret:%@, order_id:%@, orderInfo:%@, amount:%@, server:%@, username:%@", productID, appleSecret, orderID, orderInfo, amount, server, _userName);
    IAPDataRequest *iapData = [[IAPDataRequest alloc] initWithData:username andOrderId:orderID andOrderInfo:orderInfo andServerID:server andAmount:amount andAppleProductID:productID andAppleShareSecrect:appleSecret andRoleID:character andExtraInfo:extraInfo];
    
    [[NemoSDK sharedInstance] showIAP:iapData andMainView:self andIAPDelegate:self];
}
- (void) IAPInitFailed:(NSString *)message andErrorCode:(NSString *)errorCode
{
    dispatch_block_t block = ^{
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Init IAP Failed!" message:message delegate:nil cancelButtonTitle:@"Close" otherButtonTitles:nil, nil];
        [alert show];
    };
    //
    if ([NSThread isMainThread]) {
        block();
    } else {
        dispatch_sync(dispatch_get_main_queue(), block);
    }
}
- (void) IAPPurchaseFailed:(NSString *)message andErrorCode:(NSString *)errorCode
{
    dispatch_block_t block = ^{
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Purchase IAP Failed!" message:message delegate:nil cancelButtonTitle:@"Close" otherButtonTitles:nil, nil];
        [alert show];
    };
    //
    if ([NSThread isMainThread]) {
        block();
    } else {
        dispatch_sync(dispatch_get_main_queue(), block);
    }
}
- (void) IAPCompleted:(NSString *)message
{
    dispatch_block_t block = ^{
        UIAlertView *alert = [[UIAlertView alloc] initWithTitle:@"Payment Success!" message:message delegate:nil cancelButtonTitle:@"Close" otherButtonTitles:nil, nil];
        [alert show];
    };
    //
    if ([NSThread isMainThread]) {
        block();
    } else {
        dispatch_sync(dispatch_get_main_queue(), block);
    }
}

- (void)viewDidLoad
{
    NSLog(@"load view");
    [NemoSDK sharedInstance].loginDelegate = self;
    [super viewDidLoad];
    _btn_showiap.hidden = YES;
    _btn_userinfo.hidden = YES;
    _btn_signin.hidden = NO;
    _btn_logout.hidden = YES;
}

- (BOOL)shouldAutorotate
{
    return YES;
}

- (void)OrientationDidChange:(NSNotification*)notification
{
    //    CGSize size = [[UIScreen mainScreen] bounds].size;
    //    CGRect frame = self.view.frame;
    //    if(size.width == frame.size.width)
    //        return;
    
    //NSLog(@"w = %f, h = %f", size.width, size.height);
}
- (void) callTrackingFirebaseExample {
    //tracking start trial
        [[NemoSDKTracking Firebase] trackingEventOnFirebase:@"eventName" parameters:@{@"eventEventLogKey":@"eventEventLogValue"}];
        [[NemoSDKTracking Firebase] trackingScreenOnFirebase:@"screenName" screenClass:@"screenClass"];
        [[NemoSDKTracking Firebase] setUserPropertiesOnFirebase:@"userValue" forName:@"usernameName"];
    
}

//=========== Payment Apple IAP ==============//
- (IBAction) callIAP:(id)sender
{
    
}

#pragma Click PlayGame

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [self.view endEditing:YES];
}

@end
